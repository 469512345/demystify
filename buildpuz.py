import math

from itertools import combinations
from smt.base import *
from smt.utils import intsqrt

def buildCage(name, cells, dom):
    constraints = []
    for i1 in range(len(cells)):
        for i2 in range(i1 + 1, len(cells)):
            c1 = cells[i1]
            c2 = cells[i2]
            for v in dom:
                constraints.append(
                    Clause(
                        "{} and {} cannot both be {} as they are both in {}".format(
                            c1, c2, v, name
                        ),
                        [NeqVal(c1,v),NeqVal(c2,v)],
                    )
                )

    for v in dom:
        constraints.append(
            Clause(
                "Some cell in {} must be {}".format(name, v),
                [EqVal(c,v) for c in cells],
                [str(c) for c in cells]
            )
        )

    return constraints


def alldiffRowsCols(varmat):
    (x,y)=varmat.dim()
    dom = varmat.domain()

    constraints = []

    for col in range(y):
        constraints += buildCage("column {}".format(col+1), [varmat[i][col] for i in range(x)], dom)

    for row in range(x):
        constraints += buildCage("row {}".format(row+1), [varmat[row][i] for i in range(y)], dom)


    return constraints

# This requires a square matrix, of length n*n for some n
def boxConstraints(varmat):
    (x,y) = varmat.dim()

    s = intsqrt(x)

    assert x==y
    assert s*s==x

    constraints = []
    for i in range(0, s*s, s):
        for j in range(0, s*s, s):
            v = [varmat[i+x][j+y] for x in range(s) for y in range(s)]
            constraints += buildCage("the cage starting at top-left position ({},{})".format(i,j), v, varmat.domain())
    
    return constraints

def basicSudoku(varmat):
    constraints = []

    constraints += alldiffRowsCols(varmat)
    constraints += boxConstraints(varmat)

    return constraints