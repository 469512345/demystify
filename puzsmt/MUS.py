import random
import copy
import types
import random

from .utils import flatten, chainlist

from .base import EqVal, NeqVal

# This calculates Minimum Unsatisfiable Sets
# It users internals from solver, but is put in another file just for "neatness"


def MUS(solver, assume = tuple(), earlycutsize = None):
    smtassume = [solver._varlit2smtmap[l] for l in assume]

    core = solver.basicCore(chainlist(smtassume, solver._conlits))
    # Should never be satisfiable on the first pass
    assert core is not None
    if earlycutsize is not None and len(core) > earlycutsize:
        return None

    # So we can find different cores if we recall method
    random.shuffle(core)

    # First try chopping big bits off
    step = int(len(core) / 4)
    while step > 1:
        i = 0
        while step > 1 and i < len(core) - step:
            to_test = core[:i] + core[(i+step):]
            newcore = solver.basicCore(to_test)
            if newcore is not None:
                assert(len(newcore) < len(core))
                core = newcore
                i = 0
                step = int(len(newcore) / 4)
            else:
                i += step
        step = int(step / 2)

    # Final cleanup
    # We need to be prepared for things to disappear as
    # we reduce the core, so make a copy and iterate through
    # that
    corecpy = list(core)
    for lit in corecpy:
        if lit in core:
            to_test = list(core)
            to_test.remove(lit)
            newcore = solver.basicCore(to_test)
            if newcore is not None:
                core = newcore
    
    return [solver._conmap[x] for x in core if x in solver._conmap]



def findSmallestMUS(solver, puzlits):
    musdict = {}
    # First check for really tiny ones
    for p in puzlits:
        mus = MUS(solver, [p.neg()], 5)
        if mus is not None:
            musdict[p] = mus
    if len(musdict) > 0:
        return musdict
    for size in [100,1000,10000]:
        for p in puzlits:
            mus = MUS(solver, [p.neg()], size)
            if mus is not None:
                musdict[p] = mus
        if len(musdict) > 0:
            return musdict
