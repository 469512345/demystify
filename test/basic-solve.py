#!/usr/bin/env python3
import copy
import sys
import os
import logging

# Let me import puzsmt from one directory up
sys.path.append(os.path.dirname(os.path.dirname(os.path.realpath(__file__)))) 
import puzsmt
import puzsmt.base
import puzsmt.internal
import puzsmt.MUS
import puzsmt.prettyprint
import buildpuz

logging.basicConfig(level=logging.INFO)

# Make a matrix of variables (we can make more than one)
vars = puzsmt.base.VarMatrix(lambda t: (t[0]+1,t[1]+1), (9, 9), range(1,9+1))

# Build the puzzle (we can pass multiple matrices, depending on the puzzle)
puz = puzsmt.base.Puzzle([vars])
puz.addConstraints(buildpuz.basicSudoku(vars))


solver = puzsmt.internal.Solver(puz)

# Now, let's get an actual Sudoku!

#str = "600120384008459072000006005000264030070080006940003000310000050089700000502000190"
sudokustr = "093004560060003140004608309981345000347286951652070483406002890000400010029800034"

l = [int(c) for c in sudokustr]

sudoku = [l[i:i+9] for i in range(0, len(l), 9)]

print("Going to solve:")
print(sudoku)
# We need to put 'None' in places where we don't want a value (in case we want 0, we could hard-wire 0 = empty)

for i in range(9):
    for j in range(9):
        if sudoku[i][j] == 0:
            sudoku[i][j] = None

# First, we turn it into an assignment (remember technically an assignment is a list of variables, so we pass [sudoku])

sudokumodel = puz.assignmentToModel([sudoku])

fullsolution = solver.solveSingle(sudokumodel)

# Then we 'add' all the assignments that we know (this is what we can undo later with a 'pop')
for s in sudokumodel:
    solver.addLit(s)

# The 'puzlits' are all the booleans we have to solve
# Start by finding the ones which are not part of the known values
puzlits = [p for p in fullsolution if p not in sudokumodel]

# Store how hard the problem was to solve
trace = []

# Now, we need to check each one in turn to see which is 'cheapest'
while len(puzlits) > 0:
    musdict = puzsmt.MUS.findSmallestMUS(solver, puzlits)
    smallest = min([len(v) for v in musdict.values()])

    logging.info([(v,len(musdict[v])) for v in sorted(musdict.keys())])

    if smallest == 1:
        lits = [k for k in sorted(musdict.keys()) if len(musdict[k]) == 1]
        puzsmt.prettyprint.print_explanation(sys.stdout, solver, [musdict[l] for l in lits], lits)

        print("Doing", len(lits), " simple deductions ")

        for p in lits:
            solver.addLit(p)
            puzlits.remove(p)
    else:
        # Find first thing with smallest value
        p = [k for k in sorted(musdict.keys()) if len(musdict[k]) == smallest][0]
        puzsmt.prettyprint.print_explanation(sys.stdout, solver, musdict[p], [p])
        print("Smallest mus size:", smallest)
        trace.append(smallest)
        print("<p>Setting ", p," because:</p>")
        print("<ul>")
        for clause in sorted(musdict[p]):
            logging.info(clause.clauseset())
            print("<li>", solver.explain(clause), "</li>")
        print("</ul>")
        solver.addLit(p)
        puzlits.remove(p)
    
    print("<hr>")
        
print("Trace: ", trace)