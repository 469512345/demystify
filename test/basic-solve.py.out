Going to solve:
[[0, 9, 3, 0, 0, 4, 5, 6, 0], [0, 6, 0, 0, 0, 3, 1, 4, 0], [0, 0, 4, 6, 0, 8, 3, 0, 9], [9, 8, 1, 3, 4, 5, 0, 0, 0], [3, 4, 7, 2, 8, 6, 9, 5, 1], [6, 5, 2, 0, 7, 0, 4, 8, 3], [4, 0, 6, 0, 0, 2, 8, 9, 0], [0, 0, 0, 4, 0, 0, 0, 1, 0], [0, 2, 9, 8, 0, 0, 0, 3, 4]]
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 1) is not 3
Setting  (1, 1) is not 4
Setting  (1, 1) is not 5
Setting  (1, 1) is not 6
Setting  (1, 1) is not 9
Setting  (1, 2) is not 1
Setting  (1, 2) is not 2
Setting  (1, 2) is not 3
Setting  (1, 2) is not 4
Setting  (1, 2) is not 5
Setting  (1, 2) is not 6
Setting  (1, 2) is not 7
Setting  (1, 2) is not 8
Setting  (1, 3) is not 1
Setting  (1, 3) is not 2
Setting  (1, 3) is not 4
Setting  (1, 3) is not 5
Setting  (1, 3) is not 6
Setting  (1, 3) is not 7
Setting  (1, 3) is not 8
Setting  (1, 3) is not 9
Setting  (1, 4) is not 3
Setting  (1, 4) is not 4
Setting  (1, 4) is not 5
Setting  (1, 4) is not 6
Setting  (1, 4) is not 8
Setting  (1, 4) is not 9
Setting  (1, 5) is not 3
Setting  (1, 5) is not 4
Setting  (1, 5) is not 5
Setting  (1, 5) is not 6
Setting  (1, 5) is not 7
Setting  (1, 5) is not 8
Setting  (1, 5) is not 9
Setting  (1, 6) is not 1
Setting  (1, 6) is not 2
Setting  (1, 6) is not 3
Setting  (1, 6) is not 6
Setting  (1, 6) is not 7
Setting  (1, 6) is not 8
Setting  (1, 6) is not 9
Setting  (1, 7) is not 1
Setting  (1, 7) is not 2
Setting  (1, 7) is not 3
Setting  (1, 7) is not 4
Setting  (1, 7) is not 6
Setting  (1, 7) is not 7
Setting  (1, 7) is not 8
Setting  (1, 7) is not 9
Setting  (1, 8) is not 1
Setting  (1, 8) is not 2
Setting  (1, 8) is not 3
Setting  (1, 8) is not 4
Setting  (1, 8) is not 5
Setting  (1, 8) is not 7
Setting  (1, 8) is not 8
Setting  (1, 8) is not 9
Setting  (1, 9) is not 1
Setting  (1, 9) is not 3
Setting  (1, 9) is not 4
Setting  (1, 9) is not 5
Setting  (1, 9) is not 6
Setting  (1, 9) is not 9
Setting  (2, 1) is not 1
Setting  (2, 1) is not 3
Setting  (2, 1) is not 4
Setting  (2, 1) is not 6
Setting  (2, 1) is not 9
Setting  (2, 2) is not 1
Setting  (2, 2) is not 2
Setting  (2, 2) is not 3
Setting  (2, 2) is not 4
Setting  (2, 2) is not 5
Setting  (2, 2) is not 7
Setting  (2, 2) is not 8
Setting  (2, 2) is not 9
Setting  (2, 3) is not 1
Setting  (2, 3) is not 2
Setting  (2, 3) is not 3
Setting  (2, 3) is not 4
Setting  (2, 3) is not 6
Setting  (2, 3) is not 9
Setting  (2, 4) is not 1
Setting  (2, 4) is not 3
Setting  (2, 4) is not 4
Setting  (2, 4) is not 6
Setting  (2, 4) is not 8
Setting  (2, 5) is not 1
Setting  (2, 5) is not 3
Setting  (2, 5) is not 4
Setting  (2, 5) is not 6
Setting  (2, 5) is not 7
Setting  (2, 5) is not 8
Setting  (2, 6) is not 1
Setting  (2, 6) is not 2
Setting  (2, 6) is not 4
Setting  (2, 6) is not 5
Setting  (2, 6) is not 6
Setting  (2, 6) is not 7
Setting  (2, 6) is not 8
Setting  (2, 6) is not 9
Setting  (2, 7) is not 2
Setting  (2, 7) is not 3
Setting  (2, 7) is not 4
Setting  (2, 7) is not 5
Setting  (2, 7) is not 6
Setting  (2, 7) is not 7
Setting  (2, 7) is not 8
Setting  (2, 7) is not 9
Setting  (2, 8) is not 1
Setting  (2, 8) is not 2
Setting  (2, 8) is not 3
Setting  (2, 8) is not 5
Setting  (2, 8) is not 6
Setting  (2, 8) is not 7
Setting  (2, 8) is not 8
Setting  (2, 8) is not 9
Setting  (2, 9) is not 1
Setting  (2, 9) is not 3
Setting  (2, 9) is not 4
Setting  (2, 9) is not 5
Setting  (2, 9) is not 9
Setting  (3, 1) is not 3
Setting  (3, 1) is not 4
Setting  (3, 1) is not 6
Setting  (3, 1) is not 9
Setting  (3, 2) is not 2
Setting  (3, 2) is not 3
Setting  (3, 2) is not 4
Setting  (3, 2) is not 5
Setting  (3, 2) is not 6
Setting  (3, 2) is not 8
Setting  (3, 2) is not 9
Setting  (3, 3) is not 1
Setting  (3, 3) is not 2
Setting  (3, 3) is not 3
Setting  (3, 3) is not 5
Setting  (3, 3) is not 6
Setting  (3, 3) is not 7
Setting  (3, 3) is not 8
Setting  (3, 3) is not 9
Setting  (3, 4) is not 1
Setting  (3, 4) is not 2
Setting  (3, 4) is not 3
Setting  (3, 4) is not 4
Setting  (3, 4) is not 5
Setting  (3, 4) is not 7
Setting  (3, 4) is not 8
Setting  (3, 4) is not 9
Setting  (3, 5) is not 3
Setting  (3, 5) is not 4
Setting  (3, 5) is not 6
Setting  (3, 5) is not 7
Setting  (3, 5) is not 8
Setting  (3, 5) is not 9
Setting  (3, 6) is not 2
Setting  (3, 6) is not 3
Setting  (3, 6) is not 4
Setting  (3, 6) is not 6
Setting  (3, 6) is not 7
Setting  (3, 6) is not 9
Setting  (3, 7) is not 1
Setting  (3, 7) is not 4
Setting  (3, 7) is not 5
Setting  (3, 7) is not 6
Setting  (3, 7) is not 8
Setting  (3, 7) is not 9
Setting  (3, 8) is not 1
Setting  (3, 8) is not 3
Setting  (3, 8) is not 4
Setting  (3, 8) is not 5
Setting  (3, 8) is not 6
Setting  (3, 8) is not 9
Setting  (3, 9) is not 1
Setting  (3, 9) is not 2
Setting  (3, 9) is not 4
Setting  (3, 9) is not 5
Setting  (3, 9) is not 6
Setting  (3, 9) is not 7
Setting  (3, 9) is not 8
Setting  (4, 1) is not 1
Setting  (4, 1) is not 3
Setting  (4, 1) is not 4
Setting  (4, 1) is not 5
Setting  (4, 1) is not 8
Setting  (4, 2) is not 1
Setting  (4, 2) is not 2
Setting  (4, 2) is not 3
Setting  (4, 2) is not 4
Setting  (4, 2) is not 5
Setting  (4, 2) is not 6
Setting  (4, 2) is not 7
Setting  (4, 2) is not 9
Setting  (4, 3) is not 2
Setting  (4, 3) is not 3
Setting  (4, 3) is not 4
Setting  (4, 3) is not 5
Setting  (4, 3) is not 6
Setting  (4, 3) is not 7
Setting  (4, 3) is not 8
Setting  (4, 3) is not 9
Setting  (4, 4) is not 1
Setting  (4, 4) is not 2
Setting  (4, 4) is not 4
Setting  (4, 4) is not 5
Setting  (4, 4) is not 6
Setting  (4, 4) is not 7
Setting  (4, 4) is not 8
Setting  (4, 4) is not 9
Setting  (4, 5) is not 1
Setting  (4, 5) is not 2
Setting  (4, 5) is not 3
Setting  (4, 5) is not 5
Setting  (4, 5) is not 6
Setting  (4, 5) is not 7
Setting  (4, 5) is not 8
Setting  (4, 5) is not 9
Setting  (4, 6) is not 1
Setting  (4, 6) is not 2
Setting  (4, 6) is not 3
Setting  (4, 6) is not 4
Setting  (4, 6) is not 8
Setting  (4, 7) is not 1
Setting  (4, 7) is not 3
Setting  (4, 7) is not 4
Setting  (4, 7) is not 5
Setting  (4, 7) is not 8
Setting  (4, 7) is not 9
Setting  (4, 8) is not 1
Setting  (4, 8) is not 3
Setting  (4, 8) is not 4
Setting  (4, 8) is not 5
Setting  (4, 8) is not 6
Setting  (4, 8) is not 8
Setting  (4, 8) is not 9
Setting  (4, 9) is not 1
Setting  (4, 9) is not 3
Setting  (4, 9) is not 4
Setting  (4, 9) is not 5
Setting  (4, 9) is not 8
Setting  (4, 9) is not 9
Setting  (5, 1) is not 1
Setting  (5, 1) is not 2
Setting  (5, 1) is not 4
Setting  (5, 1) is not 5
Setting  (5, 1) is not 6
Setting  (5, 1) is not 8
Setting  (5, 1) is not 9
Setting  (5, 2) is not 1
Setting  (5, 2) is not 2
Setting  (5, 2) is not 3
Setting  (5, 2) is not 5
Setting  (5, 2) is not 6
Setting  (5, 2) is not 8
Setting  (5, 2) is not 9
Setting  (5, 3) is not 1
Setting  (5, 3) is not 2
Setting  (5, 3) is not 3
Setting  (5, 3) is not 4
Setting  (5, 3) is not 5
Setting  (5, 3) is not 6
Setting  (5, 3) is not 8
Setting  (5, 3) is not 9
Setting  (5, 4) is not 1
Setting  (5, 4) is not 3
Setting  (5, 4) is not 4
Setting  (5, 4) is not 5
Setting  (5, 4) is not 6
Setting  (5, 4) is not 8
Setting  (5, 4) is not 9
Setting  (5, 5) is not 1
Setting  (5, 5) is not 3
Setting  (5, 5) is not 4
Setting  (5, 5) is not 5
Setting  (5, 5) is not 9
Setting  (5, 6) is not 1
Setting  (5, 6) is not 2
Setting  (5, 6) is not 3
Setting  (5, 6) is not 4
Setting  (5, 6) is not 5
Setting  (5, 6) is not 8
Setting  (5, 6) is not 9
Setting  (5, 7) is not 1
Setting  (5, 7) is not 2
Setting  (5, 7) is not 3
Setting  (5, 7) is not 4
Setting  (5, 7) is not 5
Setting  (5, 7) is not 6
Setting  (5, 7) is not 7
Setting  (5, 7) is not 8
Setting  (5, 8) is not 1
Setting  (5, 8) is not 2
Setting  (5, 8) is not 3
Setting  (5, 8) is not 4
Setting  (5, 8) is not 6
Setting  (5, 8) is not 7
Setting  (5, 8) is not 8
Setting  (5, 8) is not 9
Setting  (5, 9) is not 2
Setting  (5, 9) is not 3
Setting  (5, 9) is not 4
Setting  (5, 9) is not 5
Setting  (5, 9) is not 6
Setting  (5, 9) is not 7
Setting  (5, 9) is not 8
Setting  (5, 9) is not 9
Setting  (6, 1) is not 1
Setting  (6, 1) is not 2
Setting  (6, 1) is not 5
Setting  (6, 1) is not 7
Setting  (6, 1) is not 8
Setting  (6, 1) is not 9
Setting  (6, 2) is not 1
Setting  (6, 2) is not 2
Setting  (6, 2) is not 6
Setting  (6, 2) is not 7
Setting  (6, 2) is not 8
Setting  (6, 2) is not 9
Setting  (6, 3) is not 1
Setting  (6, 3) is not 5
Setting  (6, 3) is not 6
Setting  (6, 3) is not 7
Setting  (6, 3) is not 8
Setting  (6, 4) is not 2
Setting  (6, 4) is not 3
Setting  (6, 4) is not 4
Setting  (6, 4) is not 5
Setting  (6, 4) is not 6
Setting  (6, 4) is not 7
Setting  (6, 5) is not 1
Setting  (6, 5) is not 2
Setting  (6, 5) is not 3
Setting  (6, 5) is not 4
Setting  (6, 5) is not 5
Setting  (6, 5) is not 6
Setting  (6, 5) is not 9
Setting  (6, 6) is not 2
Setting  (6, 6) is not 3
Setting  (6, 6) is not 4
Setting  (6, 6) is not 5
Setting  (6, 6) is not 6
Setting  (6, 6) is not 7
Setting  (6, 7) is not 1
Setting  (6, 7) is not 3
Setting  (6, 7) is not 5
Setting  (6, 7) is not 8
Setting  (6, 7) is not 9
Setting  (6, 8) is not 1
Setting  (6, 8) is not 3
Setting  (6, 8) is not 4
Setting  (6, 8) is not 5
Setting  (6, 8) is not 6
Setting  (6, 8) is not 9
Setting  (6, 9) is not 1
Setting  (6, 9) is not 4
Setting  (6, 9) is not 5
Setting  (6, 9) is not 9
Setting  (7, 1) is not 2
Setting  (7, 1) is not 3
Setting  (7, 1) is not 6
Setting  (7, 1) is not 8
Setting  (7, 2) is not 2
Setting  (7, 2) is not 5
Setting  (7, 2) is not 6
Setting  (7, 2) is not 8
Setting  (7, 2) is not 9
Setting  (7, 3) is not 1
Setting  (7, 3) is not 2
Setting  (7, 3) is not 3
Setting  (7, 3) is not 4
Setting  (7, 3) is not 5
Setting  (7, 3) is not 8
Setting  (7, 3) is not 9
Setting  (7, 4) is not 2
Setting  (7, 4) is not 3
Setting  (7, 4) is not 4
Setting  (7, 4) is not 6
Setting  (7, 4) is not 8
Setting  (7, 4) is not 9
Setting  (7, 5) is not 2
Setting  (7, 5) is not 4
Setting  (7, 5) is not 6
Setting  (7, 5) is not 7
Setting  (7, 5) is not 8
Setting  (7, 5) is not 9
Setting  (7, 6) is not 1
Setting  (7, 6) is not 3
Setting  (7, 6) is not 4
Setting  (7, 6) is not 5
Setting  (7, 6) is not 6
Setting  (7, 6) is not 7
Setting  (7, 6) is not 8
Setting  (7, 6) is not 9
Setting  (7, 7) is not 1
Setting  (7, 7) is not 2
Setting  (7, 7) is not 3
Setting  (7, 7) is not 4
Setting  (7, 7) is not 5
Setting  (7, 7) is not 6
Setting  (7, 7) is not 7
Setting  (7, 7) is not 9
Setting  (7, 8) is not 1
Setting  (7, 8) is not 2
Setting  (7, 8) is not 3
Setting  (7, 8) is not 4
Setting  (7, 8) is not 5
Setting  (7, 8) is not 6
Setting  (7, 8) is not 7
Setting  (7, 8) is not 8
Setting  (7, 9) is not 1
Setting  (7, 9) is not 2
Setting  (7, 9) is not 3
Setting  (7, 9) is not 4
Setting  (7, 9) is not 8
Setting  (7, 9) is not 9
Setting  (8, 1) is not 2
Setting  (8, 1) is not 3
Setting  (8, 1) is not 4
Setting  (8, 1) is not 6
Setting  (8, 2) is not 2
Setting  (8, 2) is not 5
Setting  (8, 2) is not 6
Setting  (8, 2) is not 8
Setting  (8, 2) is not 9
Setting  (8, 3) is not 1
Setting  (8, 3) is not 2
Setting  (8, 3) is not 3
Setting  (8, 3) is not 4
Setting  (8, 3) is not 6
Setting  (8, 3) is not 9
Setting  (8, 4) is not 2
Setting  (8, 4) is not 3
Setting  (8, 4) is not 5
Setting  (8, 4) is not 6
Setting  (8, 4) is not 7
Setting  (8, 4) is not 8
Setting  (8, 4) is not 9
Setting  (8, 5) is not 2
Setting  (8, 5) is not 4
Setting  (8, 5) is not 7
Setting  (8, 5) is not 8
Setting  (8, 6) is not 2
Setting  (8, 6) is not 3
Setting  (8, 6) is not 4
Setting  (8, 6) is not 8
Setting  (8, 7) is not 1
Setting  (8, 7) is not 3
Setting  (8, 7) is not 4
Setting  (8, 7) is not 5
Setting  (8, 7) is not 8
Setting  (8, 7) is not 9
Setting  (8, 8) is not 3
Setting  (8, 8) is not 4
Setting  (8, 8) is not 5
Setting  (8, 8) is not 6
Setting  (8, 8) is not 8
Setting  (8, 8) is not 9
Setting  (8, 9) is not 1
Setting  (8, 9) is not 3
Setting  (8, 9) is not 4
Setting  (8, 9) is not 8
Setting  (8, 9) is not 9
Setting  (9, 1) is not 2
Setting  (9, 1) is not 3
Setting  (9, 1) is not 4
Setting  (9, 1) is not 6
Setting  (9, 1) is not 8
Setting  (9, 2) is not 1
Setting  (9, 2) is not 3
Setting  (9, 2) is not 4
Setting  (9, 2) is not 5
Setting  (9, 2) is not 6
Setting  (9, 2) is not 7
Setting  (9, 2) is not 8
Setting  (9, 2) is not 9
Setting  (9, 3) is not 1
Setting  (9, 3) is not 2
Setting  (9, 3) is not 3
Setting  (9, 3) is not 4
Setting  (9, 3) is not 5
Setting  (9, 3) is not 6
Setting  (9, 3) is not 7
Setting  (9, 3) is not 8
Setting  (9, 4) is not 1
Setting  (9, 4) is not 2
Setting  (9, 4) is not 3
Setting  (9, 4) is not 4
Setting  (9, 4) is not 5
Setting  (9, 4) is not 6
Setting  (9, 4) is not 7
Setting  (9, 4) is not 9
Setting  (9, 5) is not 2
Setting  (9, 5) is not 3
Setting  (9, 5) is not 4
Setting  (9, 5) is not 7
Setting  (9, 5) is not 8
Setting  (9, 5) is not 9
Setting  (9, 6) is not 2
Setting  (9, 6) is not 3
Setting  (9, 6) is not 4
Setting  (9, 6) is not 8
Setting  (9, 6) is not 9
Setting  (9, 7) is not 1
Setting  (9, 7) is not 2
Setting  (9, 7) is not 3
Setting  (9, 7) is not 4
Setting  (9, 7) is not 5
Setting  (9, 7) is not 8
Setting  (9, 7) is not 9
Setting  (9, 8) is not 1
Setting  (9, 8) is not 2
Setting  (9, 8) is not 4
Setting  (9, 8) is not 5
Setting  (9, 8) is not 6
Setting  (9, 8) is not 7
Setting  (9, 8) is not 8
Setting  (9, 8) is not 9
Setting  (9, 9) is not 1
Setting  (9, 9) is not 2
Setting  (9, 9) is not 3
Setting  (9, 9) is not 5
Setting  (9, 9) is not 6
Setting  (9, 9) is not 7
Setting  (9, 9) is not 8
Setting  (9, 9) is not 9
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 4) is not 2
Setting  (1, 6) is not 5
Setting  (2, 3) is not 7
Setting  (2, 4) is not 2
Setting  (2, 9) is not 6
Setting  (3, 1) is not 8
Setting  (3, 6) is not 1
Setting  (3, 6) is not 5
Setting  (3, 7) is not 2
Setting  (3, 7) is not 7
Setting  (3, 8) is not 8
Setting  (3, 9) is not 3
Setting  (4, 1) is not 2
Setting  (4, 1) is not 6
Setting  (4, 1) is not 7
Setting  (4, 6) is not 6
Setting  (4, 6) is not 7
Setting  (4, 6) is not 9
Setting  (5, 1) is not 7
Setting  (5, 2) is not 7
Setting  (5, 4) is not 7
Setting  (5, 5) is not 2
Setting  (5, 5) is not 6
Setting  (5, 5) is not 7
Setting  (5, 6) is not 7
Setting  (6, 1) is not 3
Setting  (6, 1) is not 4
Setting  (6, 2) is not 3
Setting  (6, 2) is not 4
Setting  (6, 3) is not 3
Setting  (6, 3) is not 4
Setting  (6, 3) is not 9
Setting  (6, 4) is not 8
Setting  (6, 5) is not 8
Setting  (6, 6) is not 8
Setting  (7, 1) is not 1
Setting  (7, 1) is not 5
Setting  (7, 1) is not 7
Setting  (7, 1) is not 9
Setting  (7, 2) is not 4
Setting  (7, 3) is not 7
Setting  (7, 9) is not 6
Setting  (8, 1) is not 1
Setting  (8, 1) is not 9
Setting  (8, 2) is not 1
Setting  (8, 2) is not 4
Setting  (8, 3) is not 7
Setting  (8, 5) is not 1
Setting  (8, 6) is not 1
Setting  (8, 6) is not 5
Setting  (9, 1) is not 9
Setting  (9, 6) is not 5
Smallest mus size: 1
Doing some simple deductions: 
Setting  (8, 6) is not 6
Setting  (9, 6) is not 6
Smallest mus size: 3
Setting  (6, 7) is not 6
Smallest mus size: 3
Setting  (6, 9) is not 6
Smallest mus size: 3
Setting  (6, 9) is not 8
Smallest mus size: 3
Setting  (7, 4) is not 7
Smallest mus size: 4
Setting  (6, 7) is not 2
Smallest mus size: 4
Setting  (6, 7) is not 7
Smallest mus size: 4
Setting  (6, 9) is not 2
Smallest mus size: 4
Setting  (6, 9) is not 7
Smallest mus size: 5
Setting  (6, 8) is not 2
Smallest mus size: 5
Setting  (6, 8) is not 7
Smallest mus size: 5
Setting  (8, 1) is not 5
Smallest mus size: 5
Setting  (8, 8) is not 2
Smallest mus size: 5
Setting  (8, 8) is not 7
Smallest mus size: 6
Setting  (8, 5) is not 5
Smallest mus size: 8
Setting  (7, 2) is not 1
Smallest mus size: 2
Setting  (1, 1) is not 1
Smallest mus size: 3
Setting  (3, 5) is not 1
Smallest mus size: 3
Setting  (9, 6) is not 1
Smallest mus size: 2
Setting  (9, 7) is not 7
Smallest mus size: 4
Setting  (8, 5) is not 3
Smallest mus size: 10
Setting  (4, 7) is not 6
Smallest mus size: 1
Doing some simple deductions: 
Setting  (4, 9) is 6
Smallest mus size: 1
Doing some simple deductions: 
Setting  (4, 9) is not 2
Setting  (4, 9) is not 7
Setting  (8, 9) is not 6
Smallest mus size: 10
Setting  (8, 4) is not 1
Smallest mus size: 12
Setting  (3, 1) is not 1
Smallest mus size: 1
Doing some simple deductions: 
Setting  (8, 2) is 3
Setting  (9, 1) is 1
Smallest mus size: 1
Doing some simple deductions: 
Setting  (7, 2) is not 3
Setting  (8, 2) is not 7
Setting  (9, 1) is not 5
Setting  (9, 5) is not 1
Smallest mus size: 1
Doing some simple deductions: 
Setting  (3, 2) is 1
Setting  (7, 2) is 7
Setting  (7, 5) is 3
Smallest mus size: 1
Doing some simple deductions: 
Setting  (3, 2) is not 7
Setting  (7, 5) is not 1
Setting  (7, 9) is not 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (7, 4) is 1
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 4) is not 1
Setting  (6, 4) is not 1
Setting  (7, 4) is not 5
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 4) is 7
Setting  (1, 5) is 1
Setting  (6, 4) is 9
Setting  (6, 6) is 1
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 1) is not 7
Setting  (1, 5) is not 2
Setting  (1, 9) is not 7
Setting  (2, 4) is not 7
Setting  (2, 4) is not 9
Setting  (6, 6) is not 9
Smallest mus size: 1
Doing some simple deductions: 
Setting  (2, 4) is 5
Setting  (2, 5) is 9
Smallest mus size: 1
Doing some simple deductions: 
Setting  (2, 1) is not 5
Setting  (2, 3) is not 5
Setting  (2, 5) is not 2
Setting  (2, 5) is not 5
Setting  (3, 5) is not 5
Setting  (8, 5) is not 9
Smallest mus size: 1
Doing some simple deductions: 
Setting  (2, 3) is 8
Setting  (3, 1) is 5
Setting  (3, 5) is 2
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 1) is not 8
Setting  (2, 1) is not 8
Setting  (2, 9) is not 8
Setting  (8, 3) is not 8
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 1) is 2
Setting  (1, 9) is 8
Setting  (3, 8) is not 2
Smallest mus size: 1
Doing some simple deductions: 
Setting  (1, 9) is not 2
Setting  (2, 1) is not 2
Setting  (3, 1) is not 2
Smallest mus size: 1
Doing some simple deductions: 
Setting  (2, 9) is 2
Smallest mus size: 1
Doing some simple deductions: 
Setting  (2, 9) is not 7
Setting  (8, 9) is not 2
Smallest mus size: 1
Doing some simple deductions: 
Setting  (2, 1) is 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (3, 1) is not 7
Setting  (8, 1) is not 7
Setting  (9, 1) is not 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (3, 8) is 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (4, 8) is not 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (4, 7) is 7
Setting  (4, 8) is 2
Smallest mus size: 1
Doing some simple deductions: 
Setting  (8, 7) is not 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (4, 7) is not 2
Smallest mus size: 1
Doing some simple deductions: 
Setting  (7, 9) is 5
Setting  (8, 1) is 8
Setting  (8, 6) is 9
Setting  (8, 7) is 2
Setting  (8, 9) is 7
Setting  (9, 6) is 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (8, 6) is not 7
Smallest mus size: 1
Doing some simple deductions: 
Setting  (7, 5) is not 5
Setting  (8, 3) is 5
Setting  (8, 7) is not 6
Setting  (8, 9) is not 5
Smallest mus size: 1
Doing some simple deductions: 
Setting  (8, 5) is 6
Setting  (9, 5) is 5
Setting  (9, 7) is 6
Smallest mus size: 1
Doing some simple deductions: 
Setting  (9, 5) is not 6
